ENH Use `X`'s dtype for the projection in `RBFSampler`

betatim opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
contributes to #11000

<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->


#### What does this implement/fix? Explain your changes.
This makes it so that the projection used in `RBFSampler` uses `float32` when the input `X` is `float32`, otherwise it uses `float64` as before.

#### Any other comments?
One inline comment about what to do with the docstring.

betatim committed:
Use X's dtype for the projection when it is float32

betatim committed:
Update docstring to reflect dtype change

betatim committed:
Undo blank line removal

betatim commented on sklearn/kernel_approximation.py:
```diff
@@ -262,12 +262,12 @@ class RBFSampler(_ClassNamePrefixFeaturesOutMixin, TransformerMixin, BaseEstimat
     Attributes
     ----------
-    random_offset_ : ndarray of shape (n_components,), dtype=float64
+    random_offset_ : ndarray of shape (n_components,), dtype=float64 or float32
```
What is the best way to express "mostly float64, but float32 if X is float32"?

ogrisel replied:
I think it looks good like this. `numpydoc` validation seems to be green.

ogrisel reviewed (approved):
LGTM. Can you please just fix the broken doctest in the tutorial and add an entry to document the change in the changelog (`doc/whats_new/v1.2.rst`).

betatim committed:
Mention new dtype preservation in tutorial

The tutorial mentions that input is always cast to flaot64, which is no
longer true.

betatim committed:
Add RBFSampler dtype preservation to what's new

betatim commented:
The change in `tutorial.rst` is bigger than this PR. It used to state that input is always cast to `float64` but that is no longer true. I think a better change to the tutorial would be to mention the rule (of thumb) that lets a user know if or if not the dtype will be changed. However, I don't know if there is a rule like that.

Can/should we add an example where the dtype is changed (in addition to the one that is there now where the dtype is preserved)?

ogrisel commented on doc/tutorial/basic/tutorial.rst:
```diff
@@ -227,10 +228,9 @@ Unless otherwise specified, input will be cast to ``float64``::
 
-In this example, ``X`` is ``float32``, which is cast to ``float64`` by
-``fit_transform(X)``.
+In this example, ``X`` is ``float32``, and is unchanged by ``fit_transform(X)``.
```
```suggestion
In this example, ``X`` is ``float32``, and is unchanged by ``fit_transform(X)``.

Using `float32`-typed training (or testing) data is often more 
efficient than using the usual ``float64`` ``dtype``: it allows to
reduce the memory usage and sometimes also reduces processing time
by leveraging the vector instructions of the CPU. However it can
sometimes lead to numerical stability problems causing the algorithm
to be more sensitive to the scale of the values and :ref:`require
adequate preprocessing<preprocessing_scaler>`.

Keep in mind however that not all scikit-learn estimators attempt to
work in `float32` mode. For instance, some transformers will always
cast there input to `float64` and return `float64` transformed
values as a result.
```

ogrisel commented:
> Can/should we add an example where the dtype is changed (in addition to the one that is there now where the dtype is preserved)?

I don't think so because over time those remaining float64 only estimators might disappear and we don't want to have to find a new example each time we fix the one we used as an example in the tutorial.

betatim committed:
Improve tutorial story related to input dtype casting

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

betatim committed:
Language tweak

ogrisel reviewed (approved):
Still LGTM. Thanks for the latest fixes/improvements.

thomasjpfan committed:
Merge remote-tracking branch 'upstream/main' into pr/24317

glemaitre commented on sklearn/kernel_approximation.py:
```diff
@@ -264,12 +264,12 @@ class RBFSampler(_ClassNamePrefixFeaturesOutMixin, TransformerMixin, BaseEstimat
     Attributes
     ----------
-    random_offset_ : ndarray of shape (n_components,), dtype=float64
+    random_offset_ : ndarray of shape (n_components,), dtype=float64 or float32
```
```suggestion
    random_offset_ : ndarray of shape (n_components,), dtype={np.float64, np.float32}
```

glemaitre commented on sklearn/kernel_approximation.py:
```diff
@@ -264,12 +264,12 @@ class RBFSampler(_ClassNamePrefixFeaturesOutMixin, TransformerMixin, BaseEstimat
 
     random_weights_ : ndarray of shape (n_features, n_components),\
-        dtype=float64
+        dtype=float64 or float32
```
```suggestion
        dtype={np.float64, np.float32}
```

glemaitre commented on sklearn/kernel_approximation.py:
```diff
@@ -360,6 +363,12 @@ def fit(self, X, y=None):
         )
 
         self.random_offset_ = random_state.uniform(0, 2 * np.pi, size=self.n_components)
+
```
usually in `_validate_data`, we will add the following:

```python
dtype=[np.float64, np.float32]
```

glemaitre replied:
The same option should be added in `transform`.

thomasjpfan replied:
For `fit`, I think this ends up to be a memory regression if `X` is an int array.

For `transform`, I'm okay with the casting. I think `safe_sparse_dot` ends up doing it anyways.

jeremiedbb replied:
I agree with @thomasjpfan that it should only be done in tranform and not in fit

betatim replied:
Given the comment in https://github.com/scikit-learn/scikit-learn/pull/24317#discussion_r964653974 (my conclusion is "The PR is fine as it is") I am not sure what the take away from this comment thread and from https://github.com/scikit-learn/scikit-learn/pull/24317#discussion_r963019800 is.

glemaitre commented on sklearn/kernel_approximation.py:
```diff
@@ -360,6 +363,12 @@ def fit(self, X, y=None):
         self.random_offset_ = random_state.uniform(0, 2 * np.pi, size=self.n_components)
+
+        # With this we preserve the dtype of X in transform() later
+        if X.dtype == np.float32:
```
since we have the option `copy=False`, I would always add it on the line above.

glemaitre commented on sklearn/kernel_approximation.py:
```diff
@@ -360,6 +363,12 @@ def fit(self, X, y=None):
         )
 
         self.random_offset_ = random_state.uniform(0, 2 * np.pi, size=self.n_components)
```
E.g.

```suggestion
        self.random_offset_ = random_state.uniform(0, 2 * np.pi, size=self.n_components).astype(
            X.dtype, copy=False
        )
```

thomasjpfan replied:
I do not think we can do this because `X.dtype` can be an `int` here. `_validate_data` does not cast `X` into a float array.

glemaitre replied:
Not if you force the casting in validate data. Does it make sense do apply an rbf sampler on integer data?

thomasjpfan replied:
I'm mostly concern with the memory regression: https://github.com/scikit-learn/scikit-learn/pull/24317#discussion_r963045400

> Does it make sense do apply an rbf sampler on integer data?

I think it makes sense for any numerical data. The input dtype can be int, but `transform` always returns a float array.

jeremiedbb replied:
For this estimator, since fit does not use X for any computation, I think it makes sense to keep the default dtype in validate_data, i.e. ``dtype="numeric"``, not converting any numeric dtype. So we should only convert do X's dtype when it's float32, so the current version is good

betatim replied:
I was also thinking that you only want to use `float32` for `random_offset_` and `random_weights_` if `X` is of that type. And for all other types of `X` you keep the current behaviour.

glemaitre replied:
> I was also thinking that you only want to use float32 for random_offset_ and random_weights_ if X is of that type. And for all other types of X you keep the current behaviour.

I agree on the point where we should restrain to `np.float64` and `np.float32` and keep the `dtype="numeric"`.

Regarding the default dtype for lower precision than 32-bits, I am wondering if this is not counter-productive to output a 64 bits array (and store 64-bits attributes). If we head to be memory efficient, then it would make more sense to use 32 bits as a default for these lower precision dtype, isn't it?

I don't recall how much consistent we are in other estimators. This is something that we should look at before enforcing it.

betatim replied:
Can we address "32bit to save memory/speed up compute when the input isn't float64" in a different PR? I think it is a good thing to think about but it could also lead to changes in the user facing output or some other "can of worms". So maybe something for a different PR/concerted effort across all of scikit-learn?

jeremiedbb replied:
> Can we address "32bit to save memory/speed up compute when the input isn't float64" in a different PR?

I agree, let's think about that in another PR and keep the focus on just preserving float32

glemaitre reviewed (commented):
Otherwise it looks good.

betatim committed:
Fix numpy doc type annotation

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

jeremiedbb commented:
> I think that it is also worth adding a test to check that the fitted attributes are also 32 bits floating-point. This would not be covered by the common test.

And another test checking that the transformed arrays from float32 and float64 inputs are close up to some tolerance (``from sklearn.utils._testing import assert_allclose``).

betatim committed:
Add RBFSampler dtype tests

betatim commented:
I've added a test to check that the attributes are float32/float64 and that the results are `allclose`. Not sure if the `allclose` test is meaningful though, ideas for more realistic `X`s to use?

jeremiedbb commented on sklearn/kernel_approximation.py:
```diff
@@ -324,6 +324,9 @@ class RBFSampler(_ClassNamePrefixFeaturesOutMixin, TransformerMixin, BaseEstimat
         "random_state": ["random_state"],
     }
 
+    def _more_tags(self):
+        return {"preserves_dtype": [np.float64, np.float32]}
+
```
nitpick: we tend to put that at the bottom of the class

betatim replied:
Done.

@glemaitre could you update https://github.com/scikit-learn/scikit-learn/issues/11000#issue-316232750. That mentions that you should put `_more_tags` just before `__init__`.

jeremiedbb commented on sklearn/tests/test_kernel_approximation.py:
```diff
@@ -215,6 +215,33 @@ def test_rbf_sampler():
     assert np.mean(error) <= 0.05  # mean is fairly close
 
 
+@pytest.mark.parametrize("dtype", [np.float32, np.float64])
+def test_rbf_sampler_fitted_attributes_dtype(dtype):
+    # Test that attributes are float32 when input is float32
+    rbf = RBFSampler()
+
+    X = np.array([[1, 2], [3, 4], [5, 6]], dtype=dtype)
+
+    rbf.fit(X)
+
+    assert rbf.random_offset_.dtype == dtype
+    assert rbf.random_weights_.dtype == dtype
```
You can use the ``global_dtype`` fixture.
```suggestion
def test_rbf_sampler_fitted_attributes_dtype(global_dtype):
    # Test that attributes are float32 when input is float32
    rbf = RBFSampler()

    X = np.array([[1, 2], [3, 4], [5, 6]], dtype=global_dtype)

    rbf.fit(X)

    assert rbf.random_offset_.dtype == global_dtype
    assert rbf.random_weights_.dtype == global_dtype
```

betatim replied:
When I make this change the `float32` test is skipped with the following message: Set SKLEARN_RUN_FLOAT32_TESTS=1 to run float32 dtype tests.

For a test like this it feels a bit weird to skip it by default no?

jeremiedbb replied:
We set this env var in 1 of the CI jobs, so it's tested in the CI. If you want to test locally, it's true that you have to set it as well. The goal of this fixture is to avoid having the whole test suite (there's an ongoing work to improve float32 coverage in tests) run twice everytime for every job. It's already long enough.

jeremiedbb reviewed (approved):
LGTM. Just a couple of nitpicks.

betatim committed:
Move `_more_tags`

glemaitre commented on sklearn/kernel_approximation.py:
```diff
@@ -360,6 +360,12 @@ def fit(self, X, y=None):
 
         self.random_offset_ = random_state.uniform(0, 2 * np.pi, size=self.n_components)
+
+        # With this we preserve the dtype of X in transform() later
+        if X.dtype == np.float32:
```
```suggestion
        if X.dtype == np.float32:
            # Setting the data type of the fitted attribute will ensure the
            # output data type during `transform`.
```

glemaitre commented on sklearn/tests/test_kernel_approximation.py:
```diff
@@ -215,6 +215,33 @@ def test_rbf_sampler():
 
+@pytest.mark.parametrize("dtype", [np.float32, np.float64])
+def test_rbf_sampler_fitted_attributes_dtype(dtype):
+    # Test that attributes are float32 when input is float32
```
```suggestion
    """Check that the fitted attributes are stored accordingly to the
    data type of X."""
```

glemaitre commented on sklearn/tests/test_kernel_approximation.py:
```diff
@@ -215,6 +215,33 @@ def test_rbf_sampler():
+
+
+def test_rbf_sampler_dtype_equivalence():
+    # Test that the results are similar independent of X's dtype
```
```suggestion
    """Check the equivalence of the results with 32 and 64 bits input."""
```

glemaitre reviewed (approved):
LGTM once the `global_dtype` change is done.

betatim committed:
Use global_dtype fixture to avoid doubling test run length

betatim committed:
Move comments to docstrings

Co-authored-by: Guillaume Lemaitre <g.lemaitre58@gmail.com>

betatim commented:
Changed to global dtype and applied review suggestions.

I think this is ready to roll :D

Thanks for all the help and comments!

glemaitre committed:
Merge branch 'main' into dtype-RBFSampler
